File: core-presentation/src/main/java/home/oleg/placesnearme/core_presentation/recyclerview/BaseVenueAdapter.java
package home.oleg.placesnearme.core_presentation.recyclerview;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Collections;
import java.util.List;

import home.oleg.placesnearme.core_presentation.R;

/**
 * Created by Oleg Sheliakin on 23.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public abstract class BaseVenueAdapter extends ListAdapter<ItemViewType, RecyclerView.ViewHolder> {

    protected BaseVenueAdapter(@NonNull DiffUtil.ItemCallback<ItemViewType> diffCallback) {
        super(diffCallback);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        RecyclerView.ViewHolder holder;

        switch (viewType) {
            case 1:
                View venueView = LayoutInflater.from(viewGroup.getContext())
                        .inflate(getVenueItemLayoutRes(), viewGroup, false);
                holder = getVenueItemHolder(venueView);
                break;
            default:
                View emptyView = LayoutInflater.from(viewGroup.getContext())
                        .inflate(getEmptyItemLayoutRes(), viewGroup, false);
                holder = new BaseVenueAdapter.EmptyViewHolder(emptyView);
                break;
        }

        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
        ItemViewType itemViewType = getItem(i);

        if (itemViewType instanceof VenueViewItem && viewHolder instanceof BaseVenueAdapter.VenueItemHolder) {
            ((BaseVenueAdapter.VenueItemHolder) viewHolder).bind((VenueViewItem) itemViewType);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return getItem(position).getViewType();
    }

    public void showEmpty() {
        List<ItemViewType> newList = Collections.singletonList(new EmptyViewItem());
        submitList(newList);
    }

    public abstract static class VenueItemHolder extends RecyclerView.ViewHolder {
        public VenueItemHolder(@NonNull View itemView) {
            super(itemView);
        }

        protected abstract void bind(VenueViewItem venueViewItem);
    }

    private static final class EmptyViewHolder extends RecyclerView.ViewHolder {

        private EmptyViewHolder(@NonNull View itemView) {
            super(itemView);
        }
    }

    protected abstract BaseVenueAdapter.VenueItemHolder getVenueItemHolder(View view);

    @LayoutRes
    protected abstract int getVenueItemLayoutRes();

    @LayoutRes
    protected abstract int getEmptyItemLayoutRes();

}

File: core-presentation/src/main/java/home/oleg/placesnearme/core_presentation/recyclerview/BaseVenueAdapter.kt
package home.oleg.placesnearme.core_presentation.recyclerview

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.annotation.LayoutRes
import androidx.annotation.NonNull
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by Oleg Sheliakin on 23.10.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
abstract class BaseVenueAdapter
protected constructor(diffCallback: DiffUtil.ItemCallback<ItemViewType>)
    : ListAdapter<ItemViewType, RecyclerView.ViewHolder>(diffCallback) {

    @get:LayoutRes
    protected abstract val venueItemLayoutRes: Int

    @get:LayoutRes
    protected abstract val emptyItemLayoutRes: Int

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val holder: RecyclerView.ViewHolder

        when (viewType) {
            1 -> {
                val venueView = LayoutInflater.from(viewGroup.context)
                        .inflate(venueItemLayoutRes, viewGroup, false)
                holder = getVenueItemHolder(venueView)
            }
            else -> {
                val emptyView = LayoutInflater.from(viewGroup.context)
                        .inflate(emptyItemLayoutRes, viewGroup, false)
                holder = EmptyViewHolder(emptyView)
            }
        }

        return holder
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        val itemViewType = getItem(i)

        if (itemViewType is VenueViewItem && viewHolder is BaseVenueAdapter.VenueItemHolder) {
            viewHolder.bind(itemViewType)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return getItem(position).viewType
    }

    fun showEmpty() {
        //List<ItemViewType> newList = Arrays.asList(new EmptyViewItem());
        //  submitList(newList);
    }

    abstract class VenueItemHolder(@NonNull itemView: View) : RecyclerView.ViewHolder(itemView) {

        abstract fun bind(venueViewItem: VenueViewItem)
    }

    private class EmptyViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView)

    protected abstract fun getVenueItemHolder(view: View): BaseVenueAdapter.VenueItemHolder

}
